net/golioth/golioth.c

```diff
@@ -62,34 +73,84 @@ static int __golioth_close(int sock)
 	return 0;
 }
 
-static int golioth_connect_sockaddr(struct golioth_client *client, int *sock,
+static int __golioth_send(struct golioth_client *client, uint8_t *data,
+			  size_t len, int flags)
+{
+	ssize_t sent;
+
+	if (client->sock < 0) {
+		return -ENOTCONN;
+	}
+
+	sent = zsock_send(client->sock, data, len, flags);
+	if (sent < 0) {
+		return -errno;
+	} else if (sent < len) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int __golioth_send_empty_coap(struct golioth_client *client)
+{
+	struct coap_packet packet;
+	uint8_t buffer[GOLIOTH_EMPTY_PACKET_LEN];
+	int err;
+
+	err = coap_packet_init(&packet, buffer, sizeof(buffer),
+			       COAP_VERSION_1, COAP_TYPE_NON_CON,
+			       0, NULL,
+			       COAP_CODE_EMPTY, coap_next_id());
+	if (err) {
+		return err;
+	}
+
+	return __golioth_send(client, packet.data, packet.offset, 0);
+}
+
+static int golioth_connect_sockaddr(struct golioth_client *client,
 				    struct sockaddr *addr, socklen_t addrlen)
 {
+	int sock;
 	int ret;
-	int err;
+	int err = 0;
 
-	*sock = zsock_socket(addr->sa_family, SOCK_DGRAM, client->proto);
-	if (*sock < 0) {
+	sock = zsock_socket(addr->sa_family, SOCK_DGRAM, client->proto);
+	if (sock < 0) {
 		return -errno;
 	}
 
 	if (client->proto == IPPROTO_DTLS_1_2) {
-		err = golioth_setsockopt_dtls(client, *sock);
+		err = golioth_setsockopt_dtls(client, sock);
 		if (err) {
 			goto close_sock;
 		}
 	}
 
-	ret = zsock_connect(*sock, addr, addrlen);
+	ret = zsock_connect(sock, addr, addrlen);
 	if (ret < 0) {
 		err = -errno;
 		goto close_sock;
 	}
 
-	return 0;
+	golioth_lock(client);
+	client->sock = sock;
+
+	/* Send empty packet to start TLS handshake */
+	err = __golioth_send_empty_coap(client);
```

Comment:
>If I understand correctly, you're saying that the TLS handshake will occur here since it's the first send after zsock_connect. And if the handshake fails for any reason, we will detect that as a non-zero err here, correct? 

Correct.

>What happens if the handshake succeeds initially, but then the server becomes unavailable some time later. Will that case be detected and reported correctly by golioth_is_connected?

Yes, it will. There is the 9s ping interval, which will enforce some traffic before 30s RX timeout will expire. If it will expire, then `client->sock` will be set to `-1` and `golioth_is_connected()` will return `false`.